File: code-challenges/pure-code/03-dependency-resolver/solution.optimized.ts
/**
 * Dependency Resolver — Kahn’s topological sort (TypeScript)
 *
 * mode "lex": ready queue always pops lexicographically smallest name.
 * Cycle: if order.length < V after processing → cycle.
 *
 * Complejidad: O(V + E + V log V) con lex (sort/insert ready); O(V+E) sin lex.
 */

import fs from "node:fs";

interface Pkg {
  name: string;
  deps: string[];
}

type ResolveResult =
  | { ok: true; order: string[] }
  | { ok: false; error: "cycle" };

function resolve(packages: Pkg[], mode: "any" | "lex" = "lex"): ResolveResult {
  const names = packages.map((p) => p.name);
  const indegree = new Map<string, number>();
  const adj = new Map<string, string[]>(); // dep → dependents (edge dep → pkg)

  for (const name of names) {
    indegree.set(name, 0);
    adj.set(name, []);
  }

  for (const pkg of packages) {
    for (const dep of pkg.deps) {
      adj.get(dep)!.push(pkg.name);
      indegree.set(pkg.name, (indegree.get(pkg.name) ?? 0) + 1);
    }
  }

  let ready = names.filter((n) => indegree.get(n) === 0);
  if (mode === "lex") {
    ready.sort((a, b) => a.localeCompare(b));
  }

  const order: string[] = [];

  while (ready.length > 0) {
    const node = mode === "lex" ? ready.shift()! : ready.shift()!;
    order.push(node);

    for (const next of adj.get(node) ?? []) {
      const d = (indegree.get(next) ?? 0) - 1;
      indegree.set(next, d);
      if (d === 0) {
        ready.push(next);
        if (mode === "lex") {
          ready.sort((a, b) => a.localeCompare(b));
        }
      }
    }
  }

  if (order.length !== names.length) {
    return { ok: false, error: "cycle" };
  }
  return { ok: true, order };
}

function isValidOrder(packages: Pkg[], order: string[]): boolean {
  if (order.length !== packages.length) return false;
  const idx = new Map(order.map((n, i) => [n, i]));
  for (const p of packages) {
    if (!idx.has(p.name)) return false;
    for (const d of p.deps) {
      if ((idx.get(d) ?? Infinity) >= (idx.get(p.name) ?? -1)) return false;
    }
  }
  return true;
}

interface Case {
  name: string;
  mode: "any" | "lex";
  packages: Pkg[];
  expectOk: boolean;
  expectOrderIfLex?: string[];
  expectError?: string;
}

function runCase(c: Case): { name: string; passed: boolean; detail: unknown } {
  // Always lex for stable study solutions; "any" cases still validated by constraints
  const result = resolve(c.packages, "lex");
  if (!c.expectOk) {
    const passed = result.ok === false && result.error === "cycle";
    return { name: c.name, passed, detail: result };
  }

  if (!result.ok) {
    return { name: c.name, passed: false, detail: result };
  }

  let passed = isValidOrder(c.packages, result.order);
  if (c.mode === "lex" && c.expectOrderIfLex) {
    passed =
      passed &&
      result.order.length === c.expectOrderIfLex.length &&
      result.order.every((v, i) => v === c.expectOrderIfLex![i]);
  }
  return { name: c.name, passed, detail: result };
}

function main(): void {
  const path = process.argv[2] ?? "starter/cases.json";
  const raw = JSON.parse(fs.readFileSync(path, "utf8")) as { cases: Case[] };
  const results = raw.cases.map(runCase);
  const passed = results.every((r) => r.passed);
  console.log(JSON.stringify({ passed, results }, null, 2));
  process.exit(passed ? 0 : 1);
}

main();

File: code-challenges/pure-code/03-dependency-resolver/solution.optimized.py
"""
Dependency Resolver — Kahn’s topological sort (Python)

mode "lex": ready queue siempre saca el nombre lexicográficamente menor (heapq).
Cycle: si len(order) < V al terminar → cycle.

Complejidad: O((V+E) log V) con lex (heap); O(V+E) sin lex.
"""

from __future__ import annotations

import json
import sys
from collections import defaultdict
from heapq import heapify, heappop, heappush
from pathlib import Path
from typing import Any


def resolve(
    packages: list[dict[str, Any]], mode: str = "lex"
) -> dict[str, Any]:
    names = [p["name"] for p in packages]
    indegree: dict[str, int] = {n: 0 for n in names}
    adj: dict[str, list[str]] = defaultdict(list)  # dep → dependents

    for pkg in packages:
        for dep in pkg["deps"]:
            adj[dep].append(pkg["name"])
            indegree[pkg["name"]] += 1

    ready = [n for n in names if indegree[n] == 0]
    if mode == "lex":
        heapify(ready)
    order: list[str] = []

    while ready:
        node = heappop(ready) if mode == "lex" else ready.pop(0)
        order.append(node)
        for nxt in adj[node]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                if mode == "lex":
                    heappush(ready, nxt)
                else:
                    ready.append(nxt)

    if len(order) != len(names):
        return {"ok": False, "error": "cycle"}
    return {"ok": True, "order": order}


def is_valid_order(packages: list[dict], order: list[str]) -> bool:
    if len(order) != len(packages):
        return False
    idx = {n: i for i, n in enumerate(order)}
    for p in packages:
        if p["name"] not in idx:
            return False
        for d in p["deps"]:
            if idx[d] >= idx[p["name"]]:
                return False
    return True


def run_case(case: dict) -> dict:
    # Always lex for stable study solutions; "any" still validated by constraints
    result = resolve(case["packages"], "lex")

    if not case["expectOk"]:
        passed = result.get("ok") is False and result.get("error") == "cycle"
        return {"name": case["name"], "passed": passed, "detail": result}

    if not result.get("ok"):
        return {"name": case["name"], "passed": False, "detail": result}

    passed = is_valid_order(case["packages"], result["order"])
    if case.get("mode") == "lex" and case.get("expectOrderIfLex"):
        passed = passed and result["order"] == case["expectOrderIfLex"]

    return {"name": case["name"], "passed": passed, "detail": result}


def main() -> None:
    path = Path(sys.argv[1] if len(sys.argv) > 1 else "starter/cases.json")
    raw = json.loads(path.read_text(encoding="utf-8"))
    results = [run_case(c) for c in raw["cases"]]
    passed = all(r["passed"] for r in results)
    print(json.dumps({"passed": passed, "results": results}, indent=2))
    sys.exit(0 if passed else 1)


if __name__ == "__main__":
    main()
